### ISSUE TITLE
add `mask_obs`/`mask_var` arguments where appropriate

### ISSUE THREAD

**Issue by ivirshup** [scverse/scanpy#2234]
I think we should introduce a standardized “mask” argument to scanpy functions. This would be a boolean array (or reference to a boolean array in `obs`/ `var`) which masks out certain data entries.

This can be thought of as a generalization of how highly variable genes is handled. As an example:

```python
sc.pp.pca(adata, use_highly_variable=True)
```

Would be equivalent to:

```python
sc.pp.pca(adata, mask="highly_variable")
# or
sc.pp.pca(adata, mask=adata.obs["highly_variable"])
```

One of the big advantages of making this more widespread is that tasks which previously required using `.raw` or creating new anndata objects will be much easier

Some uses for this change:

### Plotting

A big one is plotting. Right now if you want to show gene expression for a subset of cells, you have to manually work with the Matplotlib Axes:

```python
ax = sc.pl.umap(pbmc, show=False)
sc.pl.umap(
    pbmc[pbmc.obs["louvain"].isin(['CD4 T cells', 'B cells', 'CD8 T cells',])],
    color="LDHB",
    ax=ax,
)
```

If a user could provide a mask, this could be reduced, and would make plotting more than one value possible:

```python
sc.pl.umap(
    pbmc,
    color=['LDHB', 'LYZ', 'CD79A’],
    mask=pbmc.obs["louvain"].isin(['CD4 T cells', 'B cells', 'CD8 T cells’,]),
)
```

### Other uses

This has come up before in a few contexts:

* Performing normalization on just some variables https://github.com/scverse/scanpy/issues/2142#issuecomment-1046729522
* Selecting a subset of variables for DE tests: https://github.com/scverse/scanpy/issues/1744
    * See also https://github.com/scverse/scanpy/issues/748
* Changing use_raw https://github.com/scverse/scanpy/issues/1798#issuecomment-819998988

## Implementation

I think this could fit quite well into the `sc.get` getter/ validation functions (https://github.com/scverse/scanpy/issues/828#issuecomment-560072919).

**Comment by flying-sheep** [comment]
@Intron7 said he had experience with this and it’s a really good way to do things fast with dask etc.

- works well in dask
- scale is a good example how to do it
- can’t do in 3rd party like PCA

**Comment by flying-sheep** [comment]
@Intron7 still waiting for your comment here!

**Comment by Intron7** [comment]
Yes, we already have a good mask for sparse scaling. Boolean arrays are very effective for indicating where computations should be performed, as they eliminate the need for copying and reintegration.

One clear example is the `tl.score_genes` function. masks there as booleans for the nanmean is a lot more efficent but less pythonic

**Comment by flying-sheep** [comment]
Where does a mask make sense:
- where is a mask better than just subsetting and then calling the function on the subset?
- where is a use case for doing it on a subset (e.g. probably none for qc, right?
- things that create an obsm/obsp entry would result in empty rows, but would there be a use case?

| Function                                                     | Should have mask | Has mask | Notes                                                        |
| ------------------------------------------------------------ | ---------------- | -------- | ------------------------------------------------------------ |
| [`pp.calculate_qc_metrics`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.calculate_qc_metrics.html) | N                | N        |                                                              |
| [`pp.filter_cells`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.filter_cells.html)/[`pp.filter_genes`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.filter_genes.html) | N                | N        | returns mask                                                 |
| [`pp.highly_variable_genes`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.highly_variable_genes.html) | N                | N        | returns mask                                                 |
| [`pp.log1p`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.log1p.html) | Maybe            | N        | because scale has it?                                        |
| [`pp.normalize_total`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.normalize_total.html) | Maybe            | N        | because scale has it?                                        |
| [`pp.regress_out`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.regress_out.html) | N                | N        |                                                              |
| [`pp.scale`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.scale.html) | Y                | Y        |                                                              |
| [`pp.subsample`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.subsample.html) | Y                | N        | maybe as `weighting`[^1]<br />also add `axis` arg            |
| [`pp.downsample_counts`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.downsample_counts.html) | N                | N        |                                                              |
| [`pp.recipe_*`](https://scanpy.readthedocs.io/en/stable/api/preprocessing.html#recipes) | N                | N        |                                                              |
| [`pp.combat`](https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.combat.html) | Y                | N        |                                                              |
| [`pp.scrublet`](https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.scrublet.html)/[`pp.scrublet_*`](https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.scrublet_simulate_doublets.html) | Maybe            | N        | maybe gene space                                             |
| [`pp.neighbors`](https://scanpy.readthedocs.io/en/stable/api/generated/scanpy.pp.neighbors.html) | Maybe            | N        | creates obsp entry<br />in relation to subset, like `.fit(X).transform(Y)` |
| [`pp.pca`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.pca.html) | Y            | Y        | has `mask_var` |
| [`tl.tnse`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.tsne.html)/[`tl.umap`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.umap.html)/[`tl.diffmap`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.diffmap.html)/[`tl.draw_graph`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.draw_graph.html#scanpy.tl.draw_graph) | Maybe            | N        | creates obsm entry<br />maybe as data source instead of `n_pcs` |
| [`tl.embedding_density`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.embedding_density.html) | N                | N        |                                                              |
| [`tl.leiden`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.leiden.html)/`tl.louvain` | Y                | Y        | as `restrict_to`[^2]                                         |
| [`tl.dendrogram` ](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.dendrogram.html) | Y                | N        | probably useful to make dendrograms for multiple subsets?    |
| [`tl.dpt`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.dpt.html) | Maybe            | N        | Ask people?                                                  |
| [`tl.paga`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.paga.html) | N                | N        |                                                              |
| [`tl.ingest`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.ingest.html) | Maybe            | N        | might be useful to use only part as reference? maybe not.    |
| [`tl.rank_genes_groups`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.rank_genes_groups.html) | N                | N        | `groups`+`reference` fullfills that purpose                  |
| [`tl.filter_rank_genes_groups`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.filter_rank_genes_groups.html) | N                | N        | creates a mask                                               |
| [`tl.marker_gene_overlap`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.marker_gene_overlap.html) | N                | N        |                                                              |
| [`tl.score_genes`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.score_genes.html) | N                | N        | uses it internally, so probably don’t expose but refactor to use it[^3] |
| [`tl.score_genes_cell_cycle`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.score_genes_cell_cycle.html) | N                | N        |                                                              |
| [`get.aggregate`](https://scanpy.readthedocs.io/en/stable/generated/scanpy.get.aggregate.html) | Y                | Y        |                                                              |

[^1]: column types: bool: subset; numeric: rel. weight per observer; cat: biased sampling
[^2]: unify `restrict_to` and `mask`?
[^3]: todo: refactor `score_genes`

**Comment by patcon** [comment]
Just wanted to say that I'm a bit fan of this proposal (`mask_obs`/`mask_var`), and going to use it on my custom clustering methods (`.tl.kmeans()` and `.tl.hdbscan()`)

I am a little curious if there's a convention for `{action}_{noun}` vs `{noun}_{action}` (vs `{action}`). I have my own preference, but I guess I'm mainly drawing attention to consistent convention :)

**Comment by flying-sheep** [comment]
due to accessors, I’m now changing this to `mask` – e.g. with shortcut syntax you’d do `mask="var.highly_variable"`.
